added nod4b

42foobar42 opened the pull request:
So here is the next mission. Happy testing...

alexander-boll committed:
added nod4b

Mailaender commented on mods/cnc/maps/nod04b/nod04b.lua:
```diff
@@ -0,0 +1,200 @@
+end
+
+ChangeOwner = function(tran, team)
+	Media.PlaySpeechNotification(player, "Reinforce")
```
This will play "Reinforcements have arrived" 3 times. Once is enough.

Mailaender commented on mods/cnc/maps/nod04b/nod04b.lua:
```diff
@@ -0,0 +1,200 @@
+
+ChangeOwner = function(tran, team)
+	Media.PlaySpeechNotification(player, "Reinforce")
+	tran.Owner = NodReinforcements
```
This looks like a crude workaround. Just make TRAN `-Selectable:`.

Mailaender commented on mods/cnc/maps/nod04b/nod04b.lua:
```diff
@@ -0,0 +1,200 @@
+	end
+end
+
+getActors = function(owner, units)
```
Please replace this with `GetActorsByType` and remove this just like you did in https://github.com/OpenRA/OpenRA/pull/7534/.

Mailaender commented:
Also the usual Z-ordering actor/tree issues you can solve by re-ordering in map.yaml:

![image](https://cloud.githubusercontent.com/assets/756669/7430069/b4cefd22-f00d-11e4-87c0-a45ef92a2014.png)

42foobar42 commented:
I could not test my changes because a get an error (out of memory) recently when I try to launch the game. Was there a PR that could cause the error (my guess is #7696).

![openra-out-of-memory](https://cloud.githubusercontent.com/assets/6122148/7436056/ee04fbaa-f049-11e4-8462-937182b15d5c.JPG)

Mailaender commented:
You are suffering from https://github.com/OpenRA/OpenRA/issues/8073. Until https://github.com/OpenRA/OpenRA/pull/7953 is reviewed and merged you will need to delete `thirdparty/SharpFont.dll` manually and `make clean`.

42foobar42 commented:
Thanks for the hint. It worked!

Mailaender commented on mods/cnc/maps/nod04b/nod04b.lua:
```diff
@@ -0,0 +1,184 @@
+
+ChangeOwner = function(tran, team)
+	
+	tran.Owner = NodReinforcements
```
I assume this workaround is in place for the `if Nod.HasNoRequiredUnits()` check. Just replace it with a bool that is set after the reinforcements are unloaded, reinforce with owner NodReinforcements and keep that while they go away.

Mailaender commented on mods/cnc/maps/nod04b/map.yaml:
```diff
@@ -0,0 +1,650 @@
+Type: Campaign
+
+Videos:
+	Briefing: nod4a.vqa
```
`nod4b.vqa` exists:

![image](https://cloud.githubusercontent.com/assets/756669/7440515/e27b4f84-f0ba-11e4-9fdc-4161c955c434.png)

42foobar42 commented:
I forget to change the video. The function ChangeOwner is not in use anymore and I deleted it.

abcdefg30 commented on mods/cnc/maps/nod04b/map.yaml:
```diff
@@ -0,0 +1,651 @@
+		NonCombatant: True
+		ColorRamp: 31,222,183
+		Race: gdi
+	PlayerReference@NodReinforcements:
```
It does not look like you're using this player anymore.
Please remove it then.

abcdefg30 commented on mods/cnc/maps/nod04b/nod04b.lua:
```diff
@@ -0,0 +1,178 @@
+			MyActors = Utils.Take(1, GDI.GetActorsByType('apc'))
+			Utils.Do(MyActors, function(actor)
+				Apc3Movement(actor)
+			end)
```
Looks like this can be simplified to
```lua
Apc3Movement(GDI.GetActorsByType('apc')[1])

42foobar42 commented:
I removed the player and simplified what you recommend.

abcdefg30 commented:
Ok, next round:
- [ ] https://github.com/OpenRA/OpenRA/pull/8048#issuecomment-98122279 needs to be resolved
- [ ] What is the `GDISupporter` player for? If you want to trigger something when a unit is discovered (`GDISupport` seems to have only one unit?), you can simply use `Trigger.OnDiscovered(Actor a, LuaFunction func)`
- [ ] You probably want to have `-Selectable:` or `Selectable: Selectable: false` on `TRAN` (See [Selectable](https://github.com/OpenRA/OpenRA/wiki/Traits#selectable))
- [ ] What about decreasing `TRAN` reveal-range a bit?
- [ ] What about a secondary objective like `Kill all GDI units in the area.`?

42foobar42 commented:
1) I hope I did not overlook something
2) Removed player and used Trigger.OnDiscovered
3) -Selectable does not work. I dont want the player give commands to the transporter.
4) decreased as much as I could
5) added secondary objective

abcdefg30 commented:
> -Selectable does not work

Yeah, indeed. Seems like a bug: #8086 
I went for 
```yaml
Selectable:
    Selectable: false
```
instead in my suggestions-pr https://github.com/42foobar42/OpenRA/pull/5.

42foobar42 commented:
I was pretty sure I tried both version and neither of them worked...
However now it does the job. Thanks!

abcdefg30 commented:
:+1: / :white_check_mark: once you squashed the commits and rebased.

42foobar42 commented:
done

obrakmann commented on mods/cnc/maps/nod04b/nod04b.lua:
```diff
@@ -0,0 +1,179 @@
+
+Apc3Movement = function(unit)
+	Utils.Do(Apc3Waypoints, function(waypoint)
+		unit.AttackMove(waypoint.Location)
```
`[string "nod04b.lua"]:61: attempt to index upvalue 'unit' (a nil value)`

abcdefg30 replied:
More like https://github.com/OpenRA/OpenRA/pull/8087#discussion_r29937395

obrakmann replied:
Yep, sorry. You're right, of course.

obrakmann commented:
:+1: This plays reasonably close to the original, and the secondary objective is appropriately hard to achieve.

The only difference is that those small huts in the south of the village needed to be destroyed in the original as well, they don't in this version. Do you want to change that, or is there a specific reasoning for that?

obrakmann commented:
Ah, you know what? It's already boring to shoot up defenseless buildings. Just adding more won't make the map any better.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/31980c4d7318aa36f08ea129ac661218596ba9fd%5E...31980c4d7318aa36f08ea129ac661218596ba9fd)

Mailaender commented:
I got a "civilian killed" when shooting the second APC.

obrakmann commented:
@Mailaender: that's correct, [that APC has civilian cargo](https://github.com/OpenRA/OpenRA/pull/8048/files#diff-fd13e7bc516499a50a52ccb474a1fb73R9).
